openedx/core/lib/cache_utils.py

```diff
@@ -6,40 +6,101 @@
+          only if you need control of your own namespaced cache - for example, to clear out your own cache.
+        - WATCH OUT: Don't use this decorator for instance methods that take in a "self" argument that changes each
+          time the method is called. This will result in constant cache misses and not provide the performance benefit
+          you are looking for. Rather, change your instance method to a class method.
+        - Benchmark, benchmark, benchmark! if you never measure, how will you know you've improved? or regressed?
 
     Arguments:
-        request_cache_attr_name - The name of the field or property in this method's containing
-         class that stores the request_cache.
+        f (func): the function to wrap
+
+    Returns:
+        func: a wrapper function which will call the wrapped function, passing in the same args/kwargs,
+              cache the value it returns, and return that cached value for subsequent calls with the
+              same args/kwargs within a single request
+    """
+    return ns_request_cached()(f)
+
+
+def ns_request_cached(namespace=None, arg_map_function=None, request_cache_getter=None):
     """
-    def _decorator(func):
-        """Outer method decorator."""
-        @functools.wraps(func)
-        def _wrapper(self, *args, **kwargs):
+    Same as request_cached above, except an optional namespace can be passed in to compartmentalize the cache.
+
+    Arguments:
+        namespace (string): An optional namespace to use for the cache.  Useful if the caller wants to manage
+            their own sub-cache by, for example, calling RequestCache(namespace=NAMESPACE).clear() for their own
+            namespace.
+        arg_map_function (function: arg->string): Function to use for mapping the wrapped function's arguments to
+            strings to use in the cache key. If not provided, defaults to force_text, which converts the given
+            argument to a string.
+        request_cache_getter (function: args->RequestCache): Function that returns the RequestCache to use. If not
+            provided, defaults to edx_django_utils.cache.RequestCache.  If None, the function's return values are
+            not cached.
+    """
+    def outer_wrapper(f):
+        """
+        Outer wrapper that decorates the given function
+
+        Arguments:
+            f (func): the function to wrap
+        """
+        def inner_wrapper(*args, **kwargs):
             """
-            Wraps a method to memoize results.
+            Wrapper function to decorate with.
             """
-            request_cache = getattr(self, request_cache_attr_name, None)
+            # Check to see if we have a result in cache.  If not, invoke our wrapped
+            # function.  Cache and return the result to the caller.
+            if request_cache_getter:
+                request_cache = request_cache_getter(args)
```

Comment:
hmm.. right now, the `has_changes` method is the only consumer of `request_cache_getter`, which doesn't need `kwargs`.  Since this is a general-purpose utility, I can add `kwargs` as well for completeness.

BTW, any better ideas of how to allow users of the decorator to pass in the request_cache they wish to use?

Reply:
updated to include `kwargs`.